pandas/core/arrays/period.py

```diff
@@ -82,7 +94,186 @@ def wrapper(self, other):
+    called "ordinals" and represent some kind of offset from a base.
+
+    The `freq` indicates the span covered by each element of the array.
+    All elements in the PeriodArray have the same `freq`.
+    """
+    _attributes = ["freq"]
+    # --------------------------------------------------------------------
+    # Constructors
+
+    def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
+                periods=None, tz=None, dtype=None, copy=False,
+                **fields):
+        from pandas import PeriodIndex, DatetimeIndex, Int64Index
+
+        # copy-pase from PeriodIndex.__new__ with slight adjustments.
+        #
+        # - removed all uses of name
+        valid_field_set = {'year', 'month', 'day', 'quarter',
+                           'hour', 'minute', 'second'}
+
+        if not set(fields).issubset(valid_field_set):
+            raise TypeError('__new__() got an unexpected keyword argument {}'.
+                            format(list(set(fields) - valid_field_set)[0]))
+
+        if periods is not None:
+            if is_float(periods):
+                periods = int(periods)
+            elif not is_integer(periods):
+                msg = 'periods must be a number, got {periods}'
+                raise TypeError(msg.format(periods=periods))
+
+        if dtype is not None:
+            dtype = pandas_dtype(dtype)
+            if not is_period_dtype(dtype):
+                raise ValueError('dtype must be PeriodDtype')
+            if freq is None:
+                freq = dtype.freq
+            elif freq != dtype.freq:
+                msg = 'specified freq and dtype are different'
+                raise IncompatibleFrequency(msg)
+
+        # coerce freq to freq object, otherwise it can be coerced elementwise
+        # which is slow
+        if freq:
+            freq = Period._maybe_convert_freq(freq)
+
+        if data is None:
+            if ordinal is not None:
+                data = np.asarray(ordinal, dtype=np.int64)
+            else:
+                data, freq = cls._generate_range(start, end, periods,
+                                                 freq, fields)
+            return cls._from_ordinals(data, freq=freq)
+
+        if isinstance(data, (PeriodArray, PeriodIndex)):
+            if freq is None or freq == data.freq:  # no freq change
+                freq = data.freq
+                data = data._ndarray_values
+            else:
+                base1, _ = _gfc(data.freq)
+                base2, _ = _gfc(freq)
+                data = libperiod.period_asfreq_arr(data._ndarray_values,
+                                                   base1, base2, 1)
+            return cls._simple_new(data, freq=freq)
+
+        # not array / index
+        if not isinstance(data, (np.ndarray, PeriodIndex,
+                                 DatetimeIndex, Int64Index)):
+            if is_scalar(data) or isinstance(data, Period):
```

Comment:
The `isinstance(data, Period)` here should be redundant.

Is scalar_data_error used anywhere else?  If not it can probably be inlined.